x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts

```diff
@@ -0,0 +1,134 @@
+
+export async function registerFleetPolicyCallbacks({
+  plugins,
+  ruleDataClient,
+  config,
+  logger,
+}: {
+  plugins: APMRouteHandlerResources['plugins'];
+  ruleDataClient: APMRouteHandlerResources['ruleDataClient'];
+  config: NonNullable<APMPlugin['currentConfig']>;
+  logger: NonNullable<APMPlugin['logger']>;
+}) {
+  if (!plugins.fleet) {
+    return;
+  }
+  const fleetPluginStart = await plugins.fleet.start();
+
+  // Registers a callback invoked when a policy is created to populate the APM
+  // integration policy with pre-existing agent configurations
+  registerAgentConfigExternalCallback({
+    fleetPluginStart,
+    callbackName: 'packagePolicyCreate',
+    plugins,
+    ruleDataClient,
+    config,
+    logger,
+  });
+
+  // Registers a callback invoked when a policy is updated to populate the APM
+  // integration policy with existing agent configurations
+  registerAgentConfigExternalCallback({
+    fleetPluginStart,
+    callbackName: 'packagePolicyUpdate',
+    plugins,
+    ruleDataClient,
+    config,
+    logger,
+  });
+}
+
+type ExternalCallbackParams = Parameters<ExternalCallback[1]>;
+type PackagePolicy = ExternalCallbackParams[0];
+type Context = ExternalCallbackParams[1];
+type Request = ExternalCallbackParams[2];
+
+function registerAgentConfigExternalCallback({
+  fleetPluginStart,
+  callbackName,
+  plugins,
+  ruleDataClient,
+  config,
+  logger,
+}: {
+  fleetPluginStart: NonNullable<APMPluginStartDependencies['fleet']>;
+  callbackName: ExternalCallback[0];
+  plugins: APMRouteHandlerResources['plugins'];
+  ruleDataClient: APMRouteHandlerResources['ruleDataClient'];
+  config: NonNullable<APMPlugin['currentConfig']>;
+  logger: NonNullable<APMPlugin['logger']>;
+}) {
+  const callbackFn: ExternalCallback[1] = async (
+    packagePolicy: PackagePolicy,
+    context: Context,
+    request: Request
+  ) => {
+    if (packagePolicy.package?.name !== 'apm') {
+      return packagePolicy;
+    }
+    const setup = await setupRequest({
+      context: context as any,
+      params: { query: { _inspect: false } },
+      core: null as any,
+      plugins,
+      request,
+      config,
+      logger,
+      ruleDataClient,
+    });
+    const agentConfigurations = await listConfigurations({ setup });
+    return getPackagePolicyWithAgentConfigurations(
+      packagePolicy,
+      agentConfigurations
+    );
+  };
+
+  fleetPluginStart.registerExternalCallback(callbackName, callbackFn);
+}
+
+const APM_SERVER = 'apm-server';
+
+// Immutable function applies the given package policy with a set of agent configurations
+export function getPackagePolicyWithAgentConfigurations(
+  packagePolicy: PackagePolicy,
+  agentConfigurations: AgentConfiguration[]
+) {
+  const [firstInput, ...restInputs] = packagePolicy.inputs;
+  const apmServerValue = firstInput?.config?.[APM_SERVER].value;
+  return {
+    ...packagePolicy,
+    inputs: [
+      {
+        ...firstInput,
+        config: {
+          [APM_SERVER]: {
```

Comment:
@ogupte Don't you need to add the others `config` here to avoid overwriting the entire config object?
```suggestion
        config: {
          ...firstInput.config,
          [APM_SERVER]: {
```